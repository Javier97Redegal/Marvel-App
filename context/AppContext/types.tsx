import { TYPES_OF_CHARACTER_LIST } from '@/constants'

export interface AppContextProps {
    appMode: TYPES_OF_CHARACTER_LIST
    setAppMode: (mode: TYPES_OF_CHARACTER_LIST) => void
}