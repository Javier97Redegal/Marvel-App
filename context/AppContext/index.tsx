'use client'

import { createContext, useState, ReactNode, useContext } from 'react'
import { AppContextProps } from './types'
import { GetCharactersResultType } from '@/utils/MarvelAPI/AutoGeneratedTypes/GetCharacters.types'

const AppContext = createContext<AppContextProps | undefined>(undefined)

export const AppProvider = ({ children }: { children: ReactNode }) => {
    const [favorites, setFavorites] = useState<GetCharactersResultType[]>([])

    const addFavorite = (character: GetCharactersResultType) => {
        setFavorites([...favorites, character])
    }

    const removeFavorite = (character: GetCharactersResultType) => {
        setFavorites(favorites.filter(({ id }) => id !== character.id))
    }

    return <>
        <AppContext.Provider value={{ favorites, addFavorite, removeFavorite }}>
            {children}
        </AppContext.Provider>
    </>
}

export const useApp = () => {
    const context = useContext(AppContext)

    if (!context) {
        throw new Error('useApp must be used within a AppProvider')
    }

    return context
}