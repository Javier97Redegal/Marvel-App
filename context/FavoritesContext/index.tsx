import { createContext, useState, ReactNode, useContext } from 'react'
import { FavoritesContextProps } from './types'
import { GetCharactersResultType } from '@/utils/MarvelAPI/AutoGeneratedTypes/GetCharacters.types'

const FavoritesContext = createContext<FavoritesContextProps | undefined>(undefined)

export const FavoritesProvider = ({ children }: { children: ReactNode }) => {
    const [search, setSearch] = useState<string>('')
    const [favorites, setFavorites] = useState<GetCharactersResultType[]>([])

    const addFavorite = (character: GetCharactersResultType) => {
        setFavorites([...favorites, character])
    }

    const removeFavorite = (character: GetCharactersResultType) => {
        setFavorites(favorites.filter(({ id }) => id !== character.id))
    }

    const favoritesFiltered = favorites.length && search.length ? favorites.filter(({ name }) => name.toLowerCase().search(search.toLowerCase()) > -1) : favorites
    const favoritesIds = favorites.map(favorite => favorite.id)

    return <>
        <FavoritesContext.Provider value={{ favorites: favoritesFiltered, favoritesIds, addFavorite, removeFavorite, searchFavorites: setSearch }}>
            {children}
        </FavoritesContext.Provider>
    </>
}

export const useFavorites = () => {
    const context = useContext(FavoritesContext)

    if (!context) {
        throw new Error('useFavorites must be used within a FavoritesProvider')
    }

    return context
}