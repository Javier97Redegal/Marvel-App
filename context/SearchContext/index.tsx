import { createContext, useState, ReactNode, useContext, useEffect } from 'react'
import { DEFAULT_SEARCH_LIMIT } from '@/constants'
import { GetCharactersResultType } from '@/utils/MarvelAPI/AutoGeneratedTypes/GetCharacters.types'
import { getCharacters } from '@/utils/MarvelAPI'
import { SearchContextProps } from './types'

const SearchContext = createContext<SearchContextProps | undefined>(undefined)

export const SearchProvider = ({ children }: { children: ReactNode }) => {
    const [search, setSearch] = useState<string>('')
    const [isLoading, setIsLoading] = useState<boolean>(true)
    const [defaultCharacters, setDefaultCharacters] = useState<GetCharactersResultType[]>([])
    const [characters, setCharacters] = useState<GetCharactersResultType[]>([])

    useEffect(() => {
        setIsLoading(true)

        if (!search && defaultCharacters.length) {
            setCharacters(defaultCharacters)
            setIsLoading(false)
            return
        }

        const fetchCharacters = async () => {
            const response = await getCharacters({
                params: {
                    limit: DEFAULT_SEARCH_LIMIT,
                    ...(!!search && { nameStartsWith: search.toLowerCase() })
                },
            })

            if (!defaultCharacters.length) {
                const defaultSearch = await getCharacters({
                    params: {
                        limit: DEFAULT_SEARCH_LIMIT,
                        ...(!!search && { nameStartsWith: search.toLowerCase() })
                    },
                })

                setDefaultCharacters(defaultSearch.data.results)
            }

            setCharacters(response.data.results)
            setIsLoading(false)
        }

        fetchCharacters()
    }, [search])

    return <>
        <SearchContext.Provider value={{ characters, searchTotal: setSearch, isLoading }}>
            {children}
        </SearchContext.Provider>
    </>
}

export const useSearch = () => {
    const context = useContext(SearchContext)

    if (!context) {
        throw new Error('useSearch must be used within a AppProvider')
    }

    return context
}