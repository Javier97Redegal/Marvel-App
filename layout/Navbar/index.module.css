.header {
    display: flex;
    justify-content: space-between;
    align-items: center;
    position: sticky;
    top: 0;
    left: 0;
    z-index: 2;
    width: 100%;
    padding: 2rem;
    background: black;
    border-bottom: 2px solid darkgray;
}

.logo {
    display: block;
}

.logo img {
    display: block;
    width: 125px;
}

.favorites {
    display: flex;
    align-items: center;
    gap: .5rem;
    font-size: 1.325rem;
    text-decoration: none;
    color: white;
}

.favorites img {
    width: 30px;
}