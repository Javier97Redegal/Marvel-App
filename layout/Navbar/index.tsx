import Link from 'next/link'
import { useApp } from '@/context/AppContext'
import { useFavorites } from '@/context/FavoritesContext'
import { TYPES_OF_CHARACTER_LIST } from '@/constants'
import styles from './index.module.css'

const Navbar = () => {
    const { setAppMode } = useApp()
    const { favoritesIds } = useFavorites()

    return <>
        <nav className={styles.header}>
            <Link onClick={() => setAppMode(TYPES_OF_CHARACTER_LIST.TOTAL)} href={{ pathname: '/' }} className={styles.logo} title='Go to start'>
                <img src={'/marvel-logo.png'} alt='Marvel Logo' />
            </Link>
            <Link onClick={() => setAppMode(TYPES_OF_CHARACTER_LIST.FAVORITES)} href={{ pathname: '/' }} className={styles.favorites} title='Go to favorites'>
                <img src={'/heart-default.png'} alt='Favorites' /> {favoritesIds.length}
            </Link>
        </nav>
    </>
}

export default Navbar