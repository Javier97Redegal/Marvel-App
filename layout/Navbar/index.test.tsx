import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import Navbar from '.'
import { useApp } from '@/context/AppContext'

jest.mock('@/context/AppContext', () => ({
    useApp: jest.fn()
}))

jest.mock('next/link', () => jest.fn(({ children, title }) => <a title={title}>{children}</a>))

describe('Navbar test', () => {
    const mockUseApp = {
        favorites: [{ id: 1 }, { id: 2 }]
    }

    beforeEach(() => {
        (useApp as jest.Mock).mockReturnValue(mockUseApp)
    })

    it('renders the navbar with the correct elements', () => {
        const { asFragment } = render(<Navbar />)

        const logoLink = screen.getByTitle('Go to start')
        const favoritesLink = screen.getByTitle('Go to favorites')

        expect(logoLink).toBeInTheDocument()
        expect(favoritesLink).toBeInTheDocument()
        expect(favoritesLink).toHaveTextContent('2')
        expect(asFragment()).toMatchSnapshot()
    })
})

