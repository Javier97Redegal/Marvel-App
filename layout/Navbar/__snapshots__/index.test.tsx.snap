// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar test renders the navbar with the correct elements 1`] = `
<DocumentFragment>
  <nav
    class="header"
  >
    <a
      title="Go to start"
    >
      <img
        alt="Marvel Logo"
        src="/marvel-logo.png"
      />
    </a>
    <a
      title="Go to favorites"
    >
      <img
        alt="Favorites"
        src="/heart-default.png"
      />
       2
    </a>
  </nav>
</DocumentFragment>
`;
